File: InJava/app/src/main/java/com/barryzeha/quizquestion/view/ScoreActivity.java
package com.barryzeha.quizquestion.view;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import com.barryzeha.quizquestion.R;
import com.barryzeha.quizquestion.databinding.ActivityScoreBinding;
import com.barryzeha.quizquestion.model.RankingModel;
import com.barryzeha.quizquestion.model.common.Constants;
import com.barryzeha.quizquestion.model.common.Util;
import com.barryzeha.quizquestion.viewModel.RankingViewModel;

import java.util.Objects;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class ScoreActivity extends AppCompatActivity {
    private RankingViewModel rankingViewModel= null;
    private int totalPoints=0;
    private ActivityScoreBinding bind;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        bind=ActivityScoreBinding.inflate(getLayoutInflater());
        Util.loadUrl(R.drawable.question_background,bind.ivScoreBackground,this);
        setContentView(bind.getRoot());

        totalPoints=getIntent().getExtras().getInt(Constants.POINTS_EXTRAS);
        setUpViewModel();
        populateScoreInfo();
        setUpListeners();
    }

    private void setUpViewModel() {
        rankingViewModel = new ViewModelProvider(this).get(RankingViewModel.class);
    }

    private void populateScoreInfo() {
        if(totalPoints!=0){
            bind.tvPoints.setText(String.valueOf(totalPoints));
           if(totalPoints >=200 && totalPoints <=300){
                bind.tvScore.setText(getString(R.string.scoreMaxMsg));
           }
           else if(totalPoints >=150 && totalPoints <=200){
               bind.tvScore.setText(getString(R.string.scoreMediumMsg));
           }
           else{
                bind.tvScore.setText(getString(R.string.lowScoreMsg));
           }
        }
    }
    private void setUpListeners(){
        bind.btnExit.setOnClickListener(v->{
            Intent intent = new Intent(this,MainActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
        });
        bind.btnSave.setOnClickListener(v->{
            if(bind.edtPlayer.getText().toString().isEmpty()){
                Toast.makeText(this, getString(R.string.writeYouName), Toast.LENGTH_SHORT).show();
            }
            else{
                saveRanking(bind.edtPlayer.getText().toString(),totalPoints);
               }
        });
    }
    private void saveRanking(String namePlayer, int points){
        try{
            Thread  thread = new Thread(() -> rankingViewModel.saveRanking(new RankingModel(namePlayer,points)));
            thread.start();
           Toast.makeText(this, getString(R.string.savedScore), Toast.LENGTH_SHORT).show();
           goToRanking();
        }
        catch(Exception e){
            e.printStackTrace();
        }
    }
    private void goToRanking(){
        bind.edtPlayer.setText("");
        bind.btnSave.setEnabled(false);
        bind.btnExit.setEnabled(false);
        Intent intent = new Intent(ScoreActivity.this, RankingActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);

        new Handler(Looper.getMainLooper()).postDelayed(() -> startActivity(intent),1000);
    }

    @Override
    public void onBackPressed() {}
}

File: InKotlin/app/src/main/java/com/barryzeha/quizquestion/view/ScoreActivity.kt
package com.barryzeha.quizquestion.view

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import androidx.activity.viewModels
import com.barryzeha.quizquestion.R
import com.barryzeha.quizquestion.common.Constant
import com.barryzeha.quizquestion.common.loadUrl
import com.barryzeha.quizquestion.common.startActivity
import com.barryzeha.quizquestion.databinding.ActivityScoreBinding
import com.barryzeha.quizquestion.model.RankingModel
import com.barryzeha.quizquestion.viewModel.RankingViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
@AndroidEntryPoint
class ScoreActivity : AppCompatActivity() {
    private lateinit var bind:ActivityScoreBinding
    private  var totalPoints:Int?=null
    private val viewModelRanking:RankingViewModel by viewModels()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bind = ActivityScoreBinding.inflate(layoutInflater)
        bind.ivScoreBackground.loadUrl(R.drawable.question_background)
        setContentView(bind.root)

        populateScoreInfo()
        setUpListeners()
    }
    private fun populateScoreInfo()= with(bind){
        totalPoints = intent?.extras?.getInt(Constant.POINTS_EXTRAS)
        totalPoints?.let{ points->
            tvPoints.text=points.toString()
            when (points) {
                in 200..300 -> {
                    tvScore.text=getString(R.string.scoreMaxMsg)
                }
                in 150 ..200 -> {
                    tvScore.text=getString(R.string.scoreMediumMsg)
                }
                else -> {
                    tvScore.text=getString(R.string.lowScoreMsg)
                }
            }
        }
    }
    private fun setUpListeners(){
        bind.btnExit.setOnClickListener {
            startActivity<MainActivity> {
                flags= Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK
            }
        }
        bind.btnSave.setOnClickListener {
            if(bind.edtPlayer.text.isNullOrEmpty()){
                Toast.makeText(this, getString(R.string.writeYouName), Toast.LENGTH_SHORT).show()
            }
            else{
                saveRanking(RankingModel(0,bind.edtPlayer.text.toString(),totalPoints!! ))
            }
        }
    }
    private fun saveRanking(ranking:RankingModel){
        CoroutineScope(Dispatchers.IO).launch {
            viewModelRanking.saveRanking(ranking)
            withContext(Dispatchers.Main){
                Toast.makeText(this@ScoreActivity, getString(R.string.savedScore), Toast.LENGTH_SHORT).show()
                goToRanking()

            }
        }
    }
    private fun goToRanking(){
        bind.edtPlayer.setText("")
        bind.btnSave.isEnabled=false
        bind.btnExit.isEnabled=false
        Handler(Looper.getMainLooper()).postDelayed({
            startActivity<RankingActivity> {
                flags= Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK
            }
        },1000)
    }
    override fun onBackPressed() {}
}